### PR TITLE
[luci/pass] Revise FoldFullyConnectedPass with luci-compute

### PR DESCRIPTION
**Pull request by seanshpark**
This will revise FoldFullyConnectedPass with luci-compute.

ONE-DCO-1.0-Signed-off-by: SaeHie Park <saehie.park@gmail.com>

### CODE REVIEW

File: compiler/luci/pass/src/FoldFullyConnectedPass.cpp [hunk 1]
```diff
@@ -173,9 +178,130 @@ bool fold_fully_connected(luci::CircleFullyConnected *node)
...
+  if (bias)
+  {
+    RETURN_FALSE_UNLESS(bias->dtype() == loco::DataType::FLOAT32);
+    bias_data = &bias->at<loco::DataType::FLOAT32>(0);
+  }
+
+  auto static_shape = [](luci::CircleNode *node) {
+    loco::TensorShape shape;
+    if (not node)
+      return shape;
```
Comment (seanshpark) [inline compiler/luci/pass/src/FoldFullyConnectedPass.cpp:261]: this is for optional bias with null

**Comment by seanshpark** [comment]
for #12100 , from  #12099


this will be the change without `#if 0 ... #endif` block

<details>
  <summary>diff</summary>

```diff
diff --git a/compiler/luci/pass/src/FoldFullyConnectedPass.cpp b/compiler/luci/pass/src/FoldFullyConnectedPass.cpp
index a3bca7eda..d67c1ea78 100644
--- a/compiler/luci/pass/src/FoldFullyConnectedPass.cpp
+++ b/compiler/luci/pass/src/FoldFullyConnectedPass.cpp
@@ -16,50 +16,47 @@

 #include "luci/Pass/FoldFullyConnectedPass.h"

-#include <tensorflow/lite/kernels/internal/reference/fully_connected.h>
+#include "helpers/Compute.h"
+#include "helpers/Shape.h"

 #include <luci/IR/CircleNodes.h>
 #include <luci/IR/AttrFusedActFunc.h>

 #include <luci/Log.h>

-#include <limits> // std::numeric_limits
+#include <luci_compute/FullyConnected.h>
+
+#include <cassert>
+
+namespace luci
+{

 namespace
 {

-bool set_kernel_parameters(tflite::FullyConnectedParams *params, luci::CircleFullyConnected *node)
+bool set_params(const luci::CircleFullyConnected *node, compute::FullyConnected &cfc)
 {
-  switch (node->fusedActivationFunction())
+  assert(node);
+
+  LOGGER(l);
+
+  // NOTE only support default for now
+  if (node->weights_format() != luci::CircleFullyConnected::WeightsFormat::DEFAULT)
   {
-    case luci::FusedActFunc::NONE:
-    case luci::FusedActFunc::TANH:
-      params->float_activation_min = std::numeric_limits<float>::lowest();
-      params->float_activation_max = std::numeric_limits<float>::max();
-      break;
-    case luci::FusedActFunc::RELU:
-      params->float_activation_min = 0;
-      params->float_activation_max = std::numeric_limits<float>::max();
-      break;
-    case luci::FusedActFunc::RELU_N1_TO_1:
-      params->float_activation_min = -1;
-      params->float_activation_max = 1;
-      break;
-    case luci::FusedActFunc::RELU6:
-      params->float_activation_min = 0;
-      params->float_activation_max = 6;
-      break;
-    default:
-    {
-      LOGGER(l);
-      WARN(l) << "Unsupported activation: " << uint32_t(node->fusedActivationFunction());
-      return false;
-    }
+    WARN(l) << "FoldFullyConnectedPass unsupported weights_format: "
+            << uint32_t(node->weights_format());
+    return false;
   }
+  cfc.params().weights_format = compute::FullyConnectedWeightsFormat::kDefault;

-  assert(node->weights_format() ==
-         luci::CircleFullyConnected::WeightsFormat::DEFAULT); // FIX_CALLER_UNLESS
-  params->weights_format = tflite::FullyConnectedWeightsFormat::kDefault;
+  compute::FusedActFunc fac;
+  if (!to_compute(node->fusedActivationFunction(), fac))
+  {
+    WARN(l) << "FoldDepthwiseConv2DPass unsupported activation: "
+            << uint32_t(node->fusedActivationFunction());
+    return false;
+  }
+  cfc.fused_act_func(fac);

   return true;
 }
@@ -85,8 +82,6 @@ bool fold_fully_connected(luci::CircleFullyConnected *node)
 {
   RETURN_FALSE_UNLESS(node != nullptr);

-  LOGGER(l);
-
   auto const input = dynamic_cast<luci::CircleConst *>(node->input());
   auto const weights = dynamic_cast<luci::CircleConst *>(node->weights());
   auto const bias = dynamic_cast<luci::CircleConst *>(node->bias());
@@ -94,77 +89,59 @@ bool fold_fully_connected(luci::CircleFullyConnected *node)

   RETURN_FALSE_UNLESS(input != nullptr);
   RETURN_FALSE_UNLESS(weights != nullptr);
-  RETURN_FALSE_UNLESS(node->weights_format() == luci::CircleFullyConnected::WeightsFormat::DEFAULT);
   RETURN_FALSE_UNLESS(bias != nullptr or no_bias != nullptr);

   RETURN_FALSE_UNLESS(input->dtype() == loco::DataType::FLOAT32);
   RETURN_FALSE_UNLESS(weights->dtype() == loco::DataType::FLOAT32);
+
+  auto const input_data = &input->at<loco::DataType::FLOAT32>(0);
+  auto const weights_data = &weights->at<loco::DataType::FLOAT32>(0);
+  float *bias_data = nullptr;
   if (bias)
+  {
     RETURN_FALSE_UNLESS(bias->dtype() == loco::DataType::FLOAT32);
+    bias_data = &bias->at<loco::DataType::FLOAT32>(0);
+  }

-  auto const input_elems = input->size<loco::DataType::FLOAT32>();
+  auto static_shape = [](luci::CircleNode *node) {
+    loco::TensorShape shape;
+    if (not node)
+      return shape;
+    shape.rank(node->rank());
+    for (uint32_t i = 0; i < node->rank(); ++i)
+      shape.dim(i) = node->dim(i);
+    return shape;
+  };

-  RETURN_FALSE_UNLESS(weights->rank() == 2);
-  RETURN_FALSE_UNLESS(input_elems % weights->dim(1).value() == 0);
-  auto const batch_size = input_elems / weights->dim(1).value();
-  auto const num_units = weights->dim(0).value();
+  compute::FullyConnected comp_fc{};
+  if (!set_params(node, comp_fc))
+    return false;
+  comp_fc.input(static_shape(input), input_data);
+  comp_fc.weights(static_shape(weights), weights_data);
+  comp_fc.bias(static_shape(bias), bias_data);

-  if (bias)
-    RETURN_FALSE_UNLESS(bias->size<loco::DataType::FLOAT32>() == num_units);
+  comp_fc.keep_num_dims(node->keep_num_dims());

-  tflite::FullyConnectedParams params{};
-  if (!set_kernel_parameters(&params, node))
-    return false; // Unsupported kernel parameter values
+  if (!comp_fc.prepare())
+    return false;

-  std::vector<uint32_t> output_shape;
-  if (node->keep_num_dims() == false)
-  {
-    output_shape.push_back(batch_size);
-    output_shape.push_back(num_units);
-  }
-  else
-  {
-    output_shape.resize(input->rank());
-    for (uint32_t i = 0; i < input->rank(); i++)
-      output_shape[i] = input->dim(i).value();
-    output_shape[input->rank() - 1] = num_units;
-  }
+  auto output_shape = comp_fc.output_shape();
+  assert(is_same_shape(node, output_shape));
+  auto output_size = loco::element_count(&output_shape);

   auto constant = node->graph()->nodes()->create<luci::CircleConst>();
   {
-    constant->name(node->name());
     constant->dtype(node->dtype());
     constant->rank(node->rank());
-    constant->shape_status(luci::ShapeStatus::VALID);
-    uint32_t num_elem = 1;
     for (uint32_t i = 0; i < node->rank(); ++i)
-    {
       constant->dim(i).set(node->dim(i).value());
-      num_elem *= node->dim(i).value();
-    }
-    constant->size<loco::DataType::FLOAT32>(num_elem);
+    constant->shape_status(luci::ShapeStatus::VALID);
+    constant->size<loco::DataType::FLOAT32>(output_size);
+    constant->name(node->name());
   }
-
-  auto tensor_shape = [](luci::CircleNode *node) {
-    if (node == nullptr)
-      return tflite::RuntimeShape();
-
-    tflite::RuntimeShape runtime_shape(node->rank());
-    for (uint32_t i = 0; i < node->rank(); ++i)
-      runtime_shape.SetDim(i, node->dim(i).value());
-    return runtime_shape;
-  };
-
-  auto tensor_data = [](luci::CircleConst *node) -> float * {
-    if (node == nullptr)
-      return nullptr;
-
-    return &node->at<loco::DataType::FLOAT32>(0);
-  };
-
-  tflite::reference_ops::FullyConnected(
-    params, tensor_shape(input), tensor_data(input), tensor_shape(weights), tensor_data(weights),
-    tensor_shape(bias), tensor_data(bias), tensor_shape(constant), tensor_data(constant));
+  auto constant_data = &constant->at<loco::DataType::FLOAT32>(0);
+  comp_fc.output(constant_data);
+  comp_fc.compute();

   loco::replace(node).with(constant);

@@ -173,9 +150,6 @@ bool fold_fully_connected(luci::CircleFullyConnected *node)

 } // namespace

-namespace luci
-{
-
 /**
  * Constant Folding for FullyConnected Op
  **/
```

```

</details>

**Review by mhs4670go — APPROVED** [review]
LGTM

File: compiler/luci/pass/src/FoldFullyConnectedPass.cpp [hunk 2]
```diff
@@ -173,9 +178,130 @@ bool fold_fully_connected(luci::CircleFullyConnected *node)
...
+    WARN(l) << "FoldFullyConnectedPass unsupported weights_format: "
+            << uint32_t(node->weights_format());
+    return false;
+  }
+  cfc.params().weights_format = compute::FullyConnectedWeightsFormat::kDefault;
+
+  compute::FusedActFunc fac;
+  if (!to_compute(node->fusedActivationFunction(), fac))
+  {
+    WARN(l) << "FoldDepthwiseConv2DPass unsupported activation: "
```
Comment (jinevening) [inline compiler/luci/pass/src/FoldFullyConnectedPass.cpp:207]:
```suggestion
    WARN(l) << "FoldFullyConnectedPass unsupported activation: "
```

Comment (seanshpark) [reply compiler/luci/pass/src/FoldFullyConnectedPass.cpp:207]: 👍🏻 👍🏻 👍🏻

**Review by jinevening — APPROVED** [review]
LGTM

**Review by mhs4670go — APPROVED** [review]

**Pull request merged**